File: Bellman Form Algorithm/C/Bellman-Ford-Algorithm.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define V 5
#define E 8

void bellmanFord(int graph[][3]) {
    int dist[V];
    for (int i = 0; i < V; i++) {
        dist[i] = INT_MAX;
    }
    dist[0] = 0;

    for (int i = 0; i < V - 1; i++) {
        for (int j = 0; j < E; j++) {
            int u = graph[j][0];
            int v = graph[j][1];
            int weight = graph[j][2];
            if (dist[u] != INT_MAX && dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight;
            }
        }
    }

    for (int i = 0; i < V; i++) {
        printf("Distance from source to vertex %d: %d\n", i, dist[i]);
    }
}

int main() {
    int graph[E][3] = {
        {0, 1, -1},
        {0, 2, 4},
        {1, 2, 3},
        {1, 3, 2},
        {1, 4, 2},
        {3, 2, 5},
        {3, 1, 1},
        {4, 3, -3}
    };

    bellmanFord(graph);

    return 0;
}

File: Bellman Form Algorithm/Python/Bellman-Ford-Algorithm.py
def bellman_ford(graph, source):
    num_vertices = len(graph)
    distances = [float('inf')] * num_vertices
    distances[source] = 0

    for _ in range(num_vertices - 1):
        for u, v, weight in graph:
            if distances[u] != float('inf') and distances[u] + weight < distances[v]:
                distances[v] = distances[u] + weight

    for u, v, weight in graph:
        if distances[u] != float('inf') and distances[u] + weight < distances[v]:
            print("Graph contains a negative-weight cycle")
            return

    print("Vertex\tDistance from Source")
    for i in range(num_vertices):
        print(f"{i}\t{distances[i]}")

graph = [
    (0, 1, -1),
    (0, 2, 4),
    (1, 2, 3),
    (1, 3, 2),
    (1, 4, 2),
    (3, 2, 5),
    (3, 1, 1),
    (4, 3, -3)
]

bellman_ford(graph, 0)
